File: app/src/main/java/dev/ragnarok/fenrir/model/IdOption.java
package dev.ragnarok.fenrir.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.api.model.Identificable;

public final class IdOption implements Parcelable, Identificable {

    public static final Creator<IdOption> CREATOR = new Creator<IdOption>() {
        @Override
        public IdOption createFromParcel(Parcel in) {
            return new IdOption(in);
        }

        @Override
        public IdOption[] newArray(int size) {
            return new IdOption[size];
        }
    };
    private final int id;
    private final String title;
    private final List<IdOption> childs;

    public IdOption(int id, String title, List<IdOption> childs) {
        this.id = id;
        this.title = title;
        this.childs = childs;
    }

    public IdOption(int id, String title) {
        this(id, title, Collections.emptyList());
    }

    private IdOption(Parcel in) {
        id = in.readInt();
        title = in.readString();
        childs = in.createTypedArrayList(CREATOR);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeString(title);
        dest.writeTypedList(childs);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public List<IdOption> getChilds() {
        return childs;
    }

    @Override
    public int getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/IdOption.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.api.model.Identificable

class IdOption : Parcelable, Identificable {
    private val id: Int
    val title: String?
    val childs: List<IdOption>?

    @JvmOverloads
    constructor(id: Int, title: String?, childs: List<IdOption>? = emptyList()) {
        this.id = id
        this.title = title
        this.childs = childs
    }

    internal constructor(parcel: Parcel) {
        id = parcel.readInt()
        title = parcel.readString()
        childs = parcel.createTypedArrayList(CREATOR)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeString(title)
        dest.writeTypedList(childs)
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun getObjectId(): Int {
        return id
    }

    companion object CREATOR : Parcelable.Creator<IdOption> {
        override fun createFromParcel(parcel: Parcel): IdOption {
            return IdOption(parcel)
        }

        override fun newArray(size: Int): Array<IdOption?> {
            return arrayOfNulls(size)
        }
    }
}